phoenix-core/src/it/java/org/apache/phoenix/end2end/OnDuplicateKey2IT.java

```diff
@@ -198,37 +199,85 @@ public void testReturnRowResult() throws Exception {
+                            + "'pk005', 23, 'col2_001')";
+            conn.createStatement().execute(upsertSql);
+
+            ps = conn.prepareStatement(
+                    "DELETE FROM " + tableName + " WHERE PK1 = ? AND PK2 = ?")
+                    .unwrap(PhoenixPreparedStatement.class);
+            ps.setString(1, "pk001");
+            ps.setDouble(2, 122.34);
+            validateReturnedRowAfterDelete(conn, ps, tableName, 2233.99, "col2_001", false,
+                    bsonDocument2, 234);
+
+            ps = conn.prepareStatement(
+                            "DELETE FROM " + tableName).unwrap(PhoenixPreparedStatement.class);
+            validateReturnedRowAfterDelete(conn, ps, tableName, 2233.99, "col2_001", false,
+                    bsonDocument2, 234);
+
+            ResultSet rs = conn.createStatement().executeQuery("SELECT * FROM " + tableName);
+            assertFalse(rs.next());
         }
     }
 
-    private static void validateReturnedRowAfterUpsert(Connection conn,
-                                                       String upsertSql,
+    private static void validateReturnedRowAfterDelete(Connection conn,
+                                                       PhoenixPreparedStatement ps,
                                                        String tableName,
                                                        Double col1,
                                                        String col2,
                                                        boolean success,
-                                                       BsonDocument inputDoc,
                                                        BsonDocument expectedDoc,
                                                        Integer col4)
             throws SQLException {
-        final Pair<Integer, Tuple> resultPair;
-        if (inputDoc != null) {
-            PhoenixPreparedStatement ps =
-                    conn.prepareStatement(upsertSql).unwrap(PhoenixPreparedStatement.class);
-            ps.setObject(1, inputDoc);
-            resultPair = ps.executeUpdateReturnRow();
-        } else {
-            resultPair = conn.createStatement().unwrap(PhoenixStatement.class)
-                    .executeUpdateReturnRow(upsertSql);
-        }
-        assertEquals(success ? 1 : 0, resultPair.getFirst().intValue());
+        final Pair<Integer, Tuple> resultPair = ps.executeUpdateReturnRow();
         ResultTuple result = (ResultTuple) resultPair.getSecond();
+
         PTable table = conn.unwrap(PhoenixConnection.class).getTable(tableName);
 
+        if (!success && result.getResult() == null) {
+            return;
+        }
         Cell cell =
                 result.getResult()
                         .getColumnLatestCell(table.getColumns().get(3).getFamilyName().getBytes(),
                                 table.getColumns().get(3).getColumnQualifierBytes());
+        if (!success) {
+            assertNull(cell);
```

Comment:
When will this happen ? The result is not null but the cell is null ?

Reply:
This happens when single key point lookup Delete is successful but it does not really delete the row because the row did not exist in the first place when it tried to delete it atomically.